src/Pager/PagerManager.php

```diff
@@ -0,0 +1,315 @@
+ * Role of the manager
+ * -------------------
+ * - manages Pager elements which are atomic objects of a paging reference and roughly corresponding current Bolt\Pager
+ *   objects
+ * - responsible for decoding/encoding pager objects from/to query parameters
+ * - centralizing all pager related operations like
+ *      ``$page = ($request->query) ? $request->query->get($param, $request->query->get('page', 1)) : 1;``
+ *      that occurs redundantly in code atm
+ * - a pager element would be reached as ``$app['pager']['search']``
+ * - ``(string) $app['pager']`` for encoding html query rather than ``Pager::makeLink()``
+ * - no more ``Bolt\Legacy\Storage::GetContent()`` (and others) has to receive &$pager as argument, which is not so clear
+ *   enough
+ *
+ * Conventions:
+ * ------------
+ *  - *Context Id* : Textual id of a pager element. It is hints the context or content type where the pager is refers to
+ *  - *Pager Id* / *Parameter Id* : Full text id of a pager object in placeholder array. It is a key with
+ * ``<PAGE>_<context_id>``. Query parameters can contain current page state under this parameter id.
+ *
+ * Practical:
+ * ----------
+ *  - Variable ``pager`` injected into templates contains a member ``manager`` furthermore, so PagerManager API can be
+ *  accessed via
+ *  - Instantiating PagerManager - like reaching ``$app['pager']`` - decodes http page parameters and can be addressed
+ *  by their context id. So ``$app['pager']['entities']`` returns a Pager object decoded from query parameters was
+ *  ``page_entities=N`` originally
+ *
+ * @package Bolt\Pager
+ * @author Rix Beck <rix@neologik.hu>
+ */
+class PagerManager implements \ArrayAccess
+{
+    const PAGE = 'page';
+
+    protected $app;
+    protected $link;
+    protected $pagers = [];
+
+    public function __construct(Application $app)
+    {
+        $this->app = $app;
+        $this->pagers = $this->decodeHttpQuery();
+    }
+
+    /**
+     * Use for calling from template to build up paginated URL link.
+     * It preserves each http query parameters except which is requested and returns a http GET query string.
+     * Last parameter appended to the end of the string so its value can be just concatenating.
+     * Against deprecated Pager::makelink() this is will build link based on initialized pagers.
+     *
+     * @param string $linkFor [optional] Id of pager the link should be built for. With empty argument passing
+     *          the link will be built for the first initialized pager object found.
+     * @return string GET query string
+     */
+    public function makelink($linkFor = '')
+    {
+        /*
+         * If link set directly that forces using it rather than build
+         */
+        if ($this->link) {
+            return $this->link;
+        }
+
+        $pagerid = $this->findPagerId($linkFor);
+        $saved = $this->pagers[$pagerid];
+        unset($this->pagers[$pagerid]);
+
+        $qparams = $this->app['request']->query->all();
```

Comment:
Can you use `$app['request_stack']->getCurrentRequest()` and check if it is `null` before using to prevent use out of cycle.